api/client_test.go

```diff
@@ -0,0 +1,274 @@
+
+	// run server using httptest
+	server := httptest.NewServer(handler)
+	defer server.Close()
+
+	var wg sync.WaitGroup
+
+	t.Run("create", func(t *testing.T) {
+		wg.Add(n)
+		id := "test"
+		var success int
+		var ltConfig loadtest.Config
+		var ucConfig simulcontroller.Config
+		defaults.Set(&ltConfig)
+		defaults.Set(&ucConfig)
+		agent, err := agentClient.New(id, server.URL, nil)
+		require.NoError(t, err)
+		require.NotNil(t, agent)
+		for i := 0; i < n; i++ {
+			go func() {
+				if _, err := agent.Create(&ltConfig, &ucConfig); err == nil {
+					success += 1
+				}
+				wg.Done()
+			}()
+		}
+		wg.Wait()
+		// Only the one attempt should have succeeded.
+		require.Equal(t, 1, success)
+
+		_, err = agent.Run()
+		require.NoError(t, err)
+		_, err = agent.AddUsers(10)
+		require.NoError(t, err)
+
+		success = 0
+		wg.Add(n)
+		for i := 0; i < n; i++ {
+			go func() {
+				_, err := agent.Destroy()
+				if err == nil {
+					success += 1
+				}
+				wg.Done()
+			}()
+		}
+		wg.Wait()
+		// Only the one attempt should have succeeded.
+		require.Equal(t, 1, success)
+	})
+
+	t.Run("create/destroy", func(t *testing.T) {
+		wg.Add(n)
+		for i := 0; i < n; i++ {
+			go func(id string) {
+				agent := createAgent(t, id, server.URL)
+				require.Equal(t, id, agent.Id())
+				_, err := agent.Destroy()
+				require.NoError(t, err)
+				wg.Done()
+			}(fmt.Sprintf("agent-%d", i))
+		}
+		wg.Wait()
+	})
+
+	t.Run("run/stop", func(t *testing.T) {
+		wg.Add(n)
+		for i := 0; i < n; i++ {
+			go func(id string) {
+				agent := createAgent(t, id, server.URL)
+				require.Equal(t, id, agent.Id())
+				st, err := agent.Run()
+				require.NoError(t, err)
+				require.Equal(t, loadtest.Running, st.State)
+				time.Sleep(100 * time.Millisecond)
+				st, err = agent.Stop()
+				require.NoError(t, err)
+				require.Equal(t, loadtest.Stopped, st.State)
+				st, err = agent.Run()
+				require.NoError(t, err)
+				require.Equal(t, loadtest.Running, st.State)
+				time.Sleep(100 * time.Millisecond)
+				st, err = agent.Stop()
+				require.NoError(t, err)
+				require.Equal(t, loadtest.Stopped, st.State)
+				wg.Done()
+			}(fmt.Sprintf("agent-%d", i))
+		}
+		wg.Wait()
+	})
+
+	t.Run("status", func(t *testing.T) {
+		id := "test"
+		agent := createAgent(t, id, server.URL)
+		require.Equal(t, id, agent.Id())
+		wg.Add(n)
+		for i := 0; i < n; i++ {
+			go func() {
+				st, err := agent.Status()
+				require.NoError(t, err)
+				require.Empty(t, st)
+				wg.Done()
+			}()
+		}
+		wg.Wait()
+		_, err := agent.Destroy()
+		require.NoError(t, err)
+	})
+
+	t.Run("add/rm users", func(t *testing.T) {
+		id := "test"
+		agent := createAgent(t, id, server.URL)
+		require.Equal(t, id, agent.Id())
+		st, err := agent.Run()
+		require.NoError(t, err)
+		require.Equal(t, loadtest.Running, st.State)
+		st, err = agent.AddUsers(n)
+		require.NoError(t, err)
+		require.Equal(t, int64(n), st.NumUsers)
+		wg.Add(n)
+		for i := 0; i < n; i++ {
+			if i%2 == 0 {
+				go func() {
```

Comment:
Wondering if we can move the goroutine a level up and do the `i%2` calculation inside it?

Reply:
If the end goal is using a single goroutine, yes, we can do that.

Reply:
0/5. This is fine too. Up to you.